v2.2/transaction-retry-errors.md

```diff
@@ -0,0 +1,152 @@
+     ```
+       SELECT x FROM a WHERE y = (SELECT q FROM b LIMIT 1)
+     ```
+     To determine the set of dependencies may require running a (sub-)query.
+   - For transactions containing more than 1 statement (interactive),
+     the set of dependencies is not computable, because the
+     transaction must start (and let some statements execute) before
+     subsequent statements start and their dependencies are known.
+2. Even if the dependencies were exactly known and it was
+   theoretically possible to compute a good ordering, this is a
+   computationally hard problem to solve (it's NP-complete) and thus
+   may not be practical to 2resolve while keeping latencies low.
+
+Because of this, most SQL engines (including CockroachDB, but also
+PostgreSQL) have a trade-off: they will **attempt** to compute
+dependencies in common cases when it is both possible and fast to do
+so, and make a **best effort** at organizing transaction ordering in a way
+that minimizes conflicts. If the dependencies cannot be computed, they
+will be pessimized (thereby assuming conflicts with a wider class of
+concurrent transactions), and if the transaction ordering risks an
+uncontrollable delay, the engine will prefer signalling the conflict
+earlier to clients.
+
+# Differences between CockroachDB and PostgreSQL
+
+How is PostgreSQL also affected? It seems like these errors never happen with PostgreSQL.
+
+- PostgreSQL [documents how and when these errors are possible](https://www.postgresql.org/docs/11/transaction-iso.html#XACT-SERIALIZABLE) at SERIALIZABLE isolation.
+- Even though it is common with other SQL engines to expect that
+  ordering errors are not possible at READ COMMITTED isolation (where
+  the **A** and **I** in ACID are weaker, in agreement with clients),
+  PostgreSQL chooses to avoid write skew effects with foreign key
+  updates (consistency, the **C** in ACID) and trigger an ordering error
+  if it detects a conflict involving foreign keys at that isolation
+  level.
+
+Why does it seem like this problem is more prevalent with CockroachDB than PostgreSQL?
+
+- The default isolation level in PostgreSQL is READ COMMITTED, not
+  SERIALIZABLE. The SERIALIZABLE level requires stricter guarantees on
+  transaction ordering and thus increases the likelihood of ordering
+  errors. This is also true in PostgreSQL, for those applications that
+  opt into that isolation level.
+
+- CockroachDB is a distributed system and it is more difficult for the
+  transaction ordering algorithms to determine the dependencies across
+  multiple nodes. The main challenge is to determine accurate
+  dependencies in a timely fashion. CockroachDB currently uses a data
+  structure that either deals with conflicts with certainty
+  (re-ordering txns automatically as needed), or there may be a conflict if two
+  transactions are close in time (the "uncertainty interval", based on
+  the MaxOffset parameter). With this implementation choice, the
```

Comment:
please put some link for the "MaxOffset parameter", otherwise readers won't know what hit them